core/src/main/java/org/infinispan/container/offheap/ExpandingOffHeapConcurrentMap.java

```diff
@@ -0,0 +1,1131 @@
+         // If the actualAddress was not known, check key equality otherwise just compare with the address
+         if (offHeapEntryFactory.equalsKey(address, key)) {
+            if (oldValue != null) {
+               ice = offHeapEntryFactory.fromMemory(address);
+               // If value doesn't match and was provided then don't replace it
+               if (!ice.getValue().equalsWrappedBytes(oldValue.getValue())) {
+                  ice = null;
+                  break;
+               }
+            }
+            // Need to always return the previous, so make sure we read it
+            if (ice == null) {
+               ice = offHeapEntryFactory.fromMemory(address);
+            }
+
+            long newAddress = offHeapEntryFactory.create(key, hashCode, newValue.getValue(), newValue.getMetadata());
+
+            entryReplaced(newAddress, address);
+            if (prevAddress != 0) {
+               offHeapEntryFactory.setNext(prevAddress, newAddress);
+            } else {
+               memoryLookup.putMemoryAddressOffset(memoryOffset, newAddress);
+            }
+            // We always set the next address on the newly created address - this will be 0 if the previous value
+            // was the end of the linked list
+            offHeapEntryFactory.setNext(newAddress, nextAddress);
+            break;
+         }
+         prevAddress = address;
+         address = nextAddress;
+      }
+      return ice;
+   }
+
+   @Override
+   public Set<WrappedBytes> keySet() {
+      throw new UnsupportedOperationException("keySet is not supported as it doesn't contain expiration data");
+   }
+
+   @Override
+   public Collection<InternalCacheEntry<WrappedBytes, WrappedBytes>> values() {
+      return new AbstractCollection<InternalCacheEntry<WrappedBytes, WrappedBytes>>() {
+         @Override
+         public Iterator<InternalCacheEntry<WrappedBytes, WrappedBytes>> iterator() {
+            return entryIterator();
+         }
+
+         @Override
+         public int size() {
+            return ExpandingOffHeapConcurrentMap.this.size();
+         }
+
+         @Override
+         public boolean remove(Object o) {
+            return o instanceof InternalCacheEntry && ExpandingOffHeapConcurrentMap.this.remove(((InternalCacheEntry) o).getKey(),
+                  ((InternalCacheEntry) o).getValue());
+         }
+      };
+   }
+
+   private class IterationPosition implements Iterator<InternalCacheEntry<WrappedBytes, WrappedBytes>> {
+      int bucketPosition;
+      int bucketCount = -1;
+      int bucketLockStop;
+      long bytesRead;
+
+      // This is arbitrary and can be changed
+      static final long READ_THRESHOLD = 8_000_000;
+      Queue<InternalCacheEntry<WrappedBytes, WrappedBytes>> values = new ArrayDeque<>();
+
+      @Override
+      public boolean hasNext() {
+         if (!values.isEmpty()) {
+            return true;
+         }
+         while (bucketPosition != bucketCount && bytesRead < READ_THRESHOLD) {
```

Comment:
Not sure I understand the question. This will read entries until either we have read all buckets or we have read bytes over the threshold and we need to wait until the entries have been returned.

Reply:
Guessing this is referring to https://github.com/infinispan/infinispan/pull/7306#discussion_r324609223

Reply:
No, I just missed the `if (!values.isEmpty())` check above :)